src/backend/metal/src/command.rs

```diff
@@ -3000,6 +3004,33 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
             if let Some(command) = self.state.set_vertex_buffers() {
                 pre.issue(command);
             }
+            // Note: all VS resources past the `old_attribute_buffer_index` have to be re-bound
+            // but ones after `pipeline.attribute_buffer_index` are already covered by `set_vertex_buffers()`.
+            // re-bind damaged VS buffers
+            let vs_end = cmp::min(self.state.resources_vs.buffers.len(), pipeline.attribute_buffer_index as usize);
+            if vs_end > old_attribute_buffer_index as usize {
+                pre.issue(soft::RenderCommand::BindBuffers {
+                    stage: pso::Stage::Vertex,
+                    index: old_attribute_buffer_index,
+                    buffers: (
+                        &self.state.resources_vs.buffers[old_attribute_buffer_index as usize .. vs_end],
+                        &self.state.resources_vs.buffer_offsets[old_attribute_buffer_index as usize .. vs_end],
+                    ),
+                });
+            }
+            // re-bind push constants
+            if let Some(index) = pipeline.vs_pc_buffer_index {
+                if Some(index) != self.state.resources_vs.push_constants_buffer_id
+                    || index >= old_attribute_buffer_index
+                {
+                    pre.issue(self.state.push_vs_constants(index));
+                }
+            }
+            if let Some(index) = pipeline.ps_pc_buffer_index {
+                if Some(index) != self.state.resources_ps.push_constants_buffer_id {
```

Comment:
Do we need to bother updating push constants when `push_constants_buffer_id` is `None`? Why not `if let Some(id) = push_constants_buffer_id` inside?

Reply:
The one in resources is the old bound buffer id. If it wasn't bound and the coming pipeline needs it, we still need to bind it, I think. `self.state.push_ps_constants(index) ` will set it to `Some`